src/pubsub/ua_pubsub_receiver.c

```diff
@@ -0,0 +1,207 @@
+#include <open62541/plugin/pubsub.h>
+#include <open62541/server_pubsub.h>
+
+#include "server/ua_server_internal.h"
+#include "ua_pubsub.h"
+#include "ua_pubsub_networkmessage.h"
+#include "ua_util_internal.h"
+
+#define MIN_PAYLOAD_SIZE_ETHERNET 46
+
+#define RECEIVE_MSG_BUFFER_SIZE   4096
+static UA_THREAD_LOCAL UA_Byte ReceiveMsgBuffer[RECEIVE_MSG_BUFFER_SIZE];
+
+/* Delete the payload value of every decoded DataSet field */
+static void UA_DataSetMessage_freeDecodedPayload(UA_DataSetMessage *dsm) {
+    if(dsm->header.fieldEncoding == UA_FIELDENCODING_VARIANT) {
+        for(UA_UInt16 i = 0; i < dsm->data.keyFrameData.fieldCount; i++) {
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+            UA_Variant_init(&dsm->data.keyFrameData.dataSetFields[i].value);
+#else
+            UA_Variant_clear(&dsm->data.keyFrameData.dataSetFields[i].value);
+#endif
+        }
+    }
+    else if(dsm->header.fieldEncoding == UA_FIELDENCODING_DATAVALUE) {
+        for(UA_UInt16 i = 0; i < dsm->data.keyFrameData.fieldCount; i++) {
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+            UA_DataValue_init(&dsm->data.keyFrameData.dataSetFields[i]);
+#else
+            UA_DataValue_clear(&dsm->data.keyFrameData.dataSetFields[i]);
+#endif
+        }
+    }
+}
+
+UA_StatusCode
+readNetworkMessage(const UA_Logger *logger,
+                               UA_ByteString *buffer, size_t *currentPosition,
+                               UA_NetworkMessage *currentNetworkMessage,
+                               UA_ReaderGroup *readerGroup) {
+
+    UA_StatusCode rv = UA_STATUSCODE_GOOD;
+#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+    rv = verifyAndDecryptNetworkMessage(logger,
+                                        buffer,
+                                        currentPosition,
+                                        currentNetworkMessage,
+                                        readerGroup);
+    UA_CHECK_STATUS_WARN(rv, return rv, logger, UA_LOGCATEGORY_SERVER,
+                         "Subscribe failed. verify and decrypt network message failed.");
+#endif
+    rv = UA_NetworkMessage_decodeBinary(buffer, currentPosition, currentNetworkMessage);
+    UA_CHECK_STATUS_WARN(rv, return rv, logger, UA_LOGCATEGORY_SERVER,
+                         "Subscribe failed. receive network message failed.");
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static
+UA_StatusCode
+receiveNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
```

Comment:
This function does not receive anything.
It processes an already received buffer.

Should be called processSomethingFooBar.